pageserver/src/page_service.rs

```diff
@@ -899,165 +995,383 @@ impl PageServerHandler {
-                            | PageStreamError::BadRequest(_) => {
-                                // print the all details to the log with {:#}, but for the client the
-                                // error message is enough.  Do not log if shutting down, as the anyhow::Error
-                                // here includes cancellation which is not an error.
-                                let full = utils::error::report_compact_sources(&e);
-                                span.in_scope(|| {
-                                    error!("error reading relation or page version: {full:#}")
-                                });
-                                PagestreamBeMessage::Error(PagestreamErrorResponse {
-                                    message: e.to_string(),
-                                })
+                        });
+                        match outcome {
+                            Outcome::Batched => {
+                                break;
                             }
-                        },
-                        Ok(response_msg) => response_msg,
-                    };
-
-                    // marshal & transmit response message
-                    pgb.write_message_noflush(&BeMessage::CopyData(&response_msg.serialize()))?;
+                            Outcome::CannotBatchNeedWaitForExecutor => {
+                                wait_notified.unwrap().await;
+                            }
+                            Outcome::ExecutorEndObserved => {
+                                break 'outer;
+                            }
+                            Outcome::Undefined => {
+                                unreachable!("send_if_modified should always be called")
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        .instrument(tracing::info_span!("batcher"));
+
+        //
+        // Create Executor future.
+        //
+
+        let executor = async {
+            let _guard = scopeguard::guard(batch_rx.clone(), |batch_rx| {
+                debug!("exiting");
+                let borrow = batch_rx.borrow();
+                let mut guard = borrow.lock().unwrap();
+                match &*guard {
+                    BatchState::Building(_) | BatchState::ReadMessagesEnded(_) => {}
+                    BatchState::ExecutorEnded => unreachable!("we only set this here"),
                 }
-                tokio::select! {
-                    biased;
-                    _ = self.cancel.cancelled() => {
-                        // We were requested to shut down.
-                        info!("shutdown request received in page handler");
-                        return Err(QueryError::Shutdown)
+                *guard = BatchState::ExecutorEnded;
+            });
+            let mut stop = false;
+            while !stop {
+                match batch_rx.changed().await {
```

Comment:
nit: does this deserve a comment mentioning that the channel has "memory" (i.e. if the change happens while we are processing things below, it still gets detected on the next call)?

Reply:
Can you check the new abstracted-out sync primitive and see whether this comment still applies? https://github.com/neondatabase/neon/pull/9851#discussion_r1859032178